File: data_structures/queues/queue.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}


class Queue {
    constructor() {
        this.first = null;
        this.last = null;
        this.length = 0;
    }

    _isEmpty() {
        return this.first == null && this.last == null;
    }

    // O(1)
    enqueue(value) {
        let newNode = new Node(value);
        if(this._isEmpty()) {
            this.first = this.last = newNode;
        }
        else {
            this.last.next = newNode;
            this.last = newNode;
        }
        this.length++;
        return this;
    }
    
    // O(1)
    dequeue() {
        if(this._isEmpty()) {
            return null;
        }
        else if (this.length === 1) {
            this.last = null;
        }
        let nodeToPop = this.first;
        this.first = nodeToPop.next;
        this.length--;
        return nodeToPop;
    }

    // O(1)
    peek() {
        return this._isEmpty()? null: this.first.value;
    }

    printAsList() {
        let array = [];
        let currentNode = this.first;
        while (currentNode !== null) {
            array.push(currentNode.value);
            currentNode = currentNode.next;
        }
        console.log(array);
    }
}

File: data_structures/queues/queue.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.first = None
        self.last = None
        self.length = 0

    def _is_empty(self):
        return self.first is None and self.last is None
    
    # O(1)
    def enqueue(self, value):
        new_node = Node(value)
        if self._is_empty():
            self.first = self.last = new_node
        else:
            self.last.next = new_node
            self.last = new_node
        self.length += 1
        return self

    # O(1)
    def dequeue(self):
        if self._is_empty():
            return None
        elif self.length == 1:
            self.last = None
        node_to_pop = self.first
        self.first = node_to_pop.next
        self.length -= 1
        return node_to_pop.value

    # O(1)
    def peek(self):
        return self.first.value if self.first is not None else None

    def print_as_list(self):
        data = []
        current_node = self.first
        while current_node is not None:
            data.append(current_node.value)
            current_node = current_node.next
        print(data)

    

    